### PR TITLE
docker image for Jupyter

### PR DESCRIPTION
**Pull request by de-code**
This makes it easier to setup the Jupyter for the notebooks. Currently the notebooks require Python 3 while the project is otherwise Python 2 (thanks to Apache Beam).

Also added the main docker image to docker compose.

(One reason for this was also to define the required dependencies for the notebooks - requirements.notebook.txt)

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        image: elifesciences/sciencebeam-judge:${IMAGE_TAG}
+    sciencebeam-judge-jupyter:
+        build:
+            context: .
+            dockerfile: Dockerfile.jupyter
+        image: elifesciences/sciencebeam-judge-jupyter:${IMAGE_TAG}
+        command: start-notebook.sh --NotebookApp.token=''
+        ports:
+            - "${SCIENCEBEAM_JUPYTER_PORT}:8888"
+        volumes:
```
Comment (giorgiosironi) [inline docker-compose.yml]: I assume the image would still be usable without the mouting of the external folder (otherwise it shouldn't be pushed as it would lack portability)

File: Jenkinsfile [hunk 2]
```diff
@@ -1,13 +1,21 @@
 elifeLibrary {
     stage 'Checkout', {
         checkout scm
+        commit = elifeGitRevision()
     }
 
-    stage 'Build image', {
-        sh 'docker build -t elife/sciencebeam-judge .'
-    }
+    node('containers-jenkins-plugin') {
```
Comment (giorgiosironi) [inline Jenkinsfile]: your can add Groovy to your CV by now

File: docker/docker-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+set -e
+
+# woraround for permissions issue (to allow notebooks to be updated and added from the container)
+
+chmod -R a+w .
```
Comment (giorgiosironi) [inline docker/docker-entrypoint.sh]: is this related to mounting the external volume? It smells a bit as the container is acting on the host filesystem, and must be run as `root` for this to work

File: Dockerfile.jupyter [hunk 4]
```diff
@@ -0,0 +1,16 @@
+FROM jupyter/minimal-notebook
+
+ENV PROJECT_HOME=/home/jovyan/sciencebeam-judge
+
+WORKDIR ${PROJECT_HOME}
+
+COPY requirements.notebook.txt ${PROJECT_HOME}/
+RUN pip install --upgrade pip && \
+  pip install -r requirements.notebook.txt
```
[the remark this answers is not in the document]
Comment (de-code) [reply Dockerfile.jupyter]: It would allow it to cache it. But in this case `--upgrade pip` doesn't specify a version. This way it will at least use the latest version when installing new dependencies. (It's also why I would put `apt update` together with `apt install`)

File: Jenkinsfile [hunk 5]
```diff
@@ -1,13 +1,21 @@
...
 
-    stage 'Run tests', {
-        elifeLocalTests './project_tests.sh'
+        stage 'Project tests', {
+            try {
+                sh "IMAGE_TAG=${commit} docker-compose run --rm sciencebeam-judge ./project_tests.sh"
+            } finally {
+                sh 'docker-compose down -v'
+            }
+        }
```
[the remark this answers is not in the document]
Comment (de-code) [reply Jenkinsfile]: It might be a good idea to be able to run it without building it. But yes, perhaps a separate PR.

File: docker-compose.yml (hunk 1 above)
Comment (de-code) [reply docker-compose.yml]: It would be useful to some extend. But I suspect not many people would want to run notebooks and edit notebooks while not also wanting to save them. If someone wants to just see the results with the example data, then they can see it within GitHub. So currently I would assume everyone who uses that container will also want to mount the directory.

File: docker/docker-entrypoint.sh (hunk 3 above)
Comment (de-code) [reply docker/docker-entrypoint.sh]: Yes, it is. Using Jupyter people can run, edit and add notebooks. The notebooks are written to disk. And by mounting them, they can be updated in git as usual.

The base image has some code in it to not run it as root. I could probably bypass it but not sure that would be better (in particular then the host needs to have root permissions to remove the files).

Do you have any other recommendation?

File: Jenkinsfile (hunk 2 above)
Comment (de-code) [reply Jenkinsfile]: Erm, I will consider it..

File: docker-compose.yml (hunk 1 above)
Comment (de-code) [reply docker-compose.yml]: Just to clarify, the image is really designed to be used with the example notebooks but that may be modified to explore other datasets.

File: docker/docker-entrypoint.sh (hunk 3 above)
Comment (giorgiosironi) [reply docker/docker-entrypoint.sh]: Good not to run as `root` if possible, like for any other application. You may want an `umask` here if it makes it easier to provide wider permissions to the files being created by the container. In general, for reuse/portability the image should try to allow running as a user passed from outside (`docker run -u`) but it takes some effort to make this possible.

Just wondering that this `chmod` could easily fail depending on the situation on the host, like files belonging to `a` and the container running as `b`.

Comment (de-code) [reply docker/docker-entrypoint.sh]: Added `umask +rw`

Comment (giorgiosironi) [reply docker/docker-entrypoint.sh]: The image should support that user if you want to give it a name but any UID is fine as far as running a process is concerned:
```
$ docker run -it -u 10001 ubuntu:18.04 /bin/bash
I have no name!@d03d63ded41c:/$
```
